File: 735-asteroid-collision/asteroid-collision.cpp
class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroid) {
        
vector<int>res;
        stack<int>st;
        for(int i=0;i<asteroid.size();i++)
        {
           if(asteroid[i]>0)
           st.push(asteroid[i]);
            else
            {
           while(!st.empty() &&st.top()>0 &&st.top()<abs(asteroid[i]))
           {
            st.pop();
           }

           if(!st.empty() &&st.top()>abs(asteroid[i]))
           {
                continue;
           }

           if(!st.empty() &&st.top()==abs(asteroid[i]))
           {
            st.pop();
            continue;
           }

           if(st.empty() ||st.top()<0)
           st.push(asteroid[i]);


            }
        }

        while(!st.empty())
        {
            res.push_back(st.top());

            st.pop();
        }
        reverse(res.begin(),res.end());

        return res;
    }
};

File: 735-asteroid-collision/asteroid-collision.py
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        st=[]
        n=len(asteroids)
        for sz in asteroids:
            if len(st)==0:
                st.append(sz)
                continue
            while st and st[-1]>0 and sz<0:
                s=abs(st[-1])
                sz_abs=abs(sz)
                if s<sz_abs:
                    st.pop()
                    
                elif sz_abs<s:
                    sz=0
                    break
                else:
                    sz=0
                    st.pop()
                    break
            if sz!=0:
                st.append(sz)
        return st

        